uqcsbot/hoogle.py

```diff
@@ -0,0 +1,64 @@
+from typing import Optional
+
+import discord
+from discord import app_commands
+from discord.ext import commands
+
+import requests
+import json
+import html
+import re
+
+class Hoogle(commands.Cog):
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+    
+    def get_endpoint(self, type_sig: str) -> str:
+        unescaped = html.unescape(type_sig)
+        return "https://www.haskell.org/hoogle/?mode=json&hoogle=" + unescaped + "&start=0&count=10"
+    
+    def get_hoogle_page(self, type_sig: str) -> str:
+        return f"https://hoogle.haskell.org/?hoogle={html.unescape(type_sig)}"
```

Comment:
Just ran into an error with these - you'll need to re-URL-encode after HTML-decoding, otherwise you can get something like `https://hoogle.haskell.org/?hoogle=/this/url/is/now/invalid`.

Reply:
For the actual haskell request that isn't a massive problem, because it'll just fail, but the link you stick in your embed gets checked for validity by discord, so it'll crash the cog.

All you need to do is `import urllib.parse` and then use `urllib.parse.quote_plus()` on the parameter before you stick it in the URL (don't encode the whole URL!).

Reply:
Oops, forgot to test what happens when there's a space in the input...
Used the quote_plus() as you mentioned as it seems to be fine now.